File: sabot/kernel/src/main/java/com/dremio/exec/physical/config/Project.java
/*
 * Copyright (C) 2017-2019 Dremio Corporation
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.dremio.exec.physical.config;

import java.util.List;

import com.dremio.common.logical.data.NamedExpression;
import com.dremio.exec.physical.base.AbstractSingle;
import com.dremio.exec.physical.base.OpProps;
import com.dremio.exec.physical.base.PhysicalOperator;
import com.dremio.exec.physical.base.PhysicalVisitor;
import com.dremio.exec.proto.UserBitShared.CoreOperatorType;
import com.fasterxml.jackson.annotation.JsonCreator;
import com.fasterxml.jackson.annotation.JsonProperty;
import com.fasterxml.jackson.annotation.JsonTypeName;
import com.google.common.collect.ImmutableList;

@JsonTypeName("project")
public class Project extends AbstractSingle {

  private final List<NamedExpression> exprs;

  @JsonCreator
  public Project(
      @JsonProperty("props") OpProps props,
      @JsonProperty("child") PhysicalOperator child,
      @JsonProperty("exprs") List<NamedExpression> exprs
      ) {
    super(props, child);
    this.exprs = (exprs == null) ? null : ImmutableList.copyOf(exprs);
  }

  /**
   * A strange current behavior here is that an empty exprs means convert to json for complex types.
   */
  public List<NamedExpression> getExprs() {
    return exprs;
  }

  @Override
  public <T, X, E extends Throwable> T accept(PhysicalVisitor<T, X, E> physicalVisitor, X value) throws E{
    return physicalVisitor.visitProject(this, value);
  }

  @Override
  protected PhysicalOperator getNewWithChild(PhysicalOperator child) {
    return new Project(props, child, exprs);
  }

  @Override
  public int getOperatorType() {
    return CoreOperatorType.PROJECT_VALUE;
  }
}

File: ui/dremio-js/src/cloud/projects/Project.ts
/*
 * Copyright (C) 2017-2019 Dremio Corporation
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import type {
  Project as ProjectInterface,
  ProjectProperties,
} from "../../interfaces/Project.js";

export class Project implements ProjectInterface {
  readonly cloudId: ProjectInterface["cloudId"];
  readonly cloudType: ProjectInterface["cloudType"];
  readonly createdAt: ProjectInterface["createdAt"];
  readonly createdBy: ProjectInterface["createdBy"];
  readonly credentials: ProjectInterface["credentials"];
  readonly id: ProjectInterface["id"];
  readonly lastStateError: ProjectInterface["lastStateError"];
  readonly modifiedAt: ProjectInterface["modifiedAt"];
  readonly modifiedBy: ProjectInterface["modifiedBy"];
  readonly name: ProjectInterface["name"];
  readonly numberOfEngines: ProjectInterface["numberOfEngines"];
  readonly primaryCatalog: ProjectInterface["primaryCatalog"];
  readonly projectStore: ProjectInterface["projectStore"];
  readonly state: ProjectInterface["state"];
  readonly type: ProjectInterface["type"];

  #apiMethods: any;

  constructor(properties: ProjectProperties, apiMethods: any) {
    this.cloudId = properties.cloudId;
    this.cloudType = properties.cloudType;
    this.createdAt = properties.createdAt;
    this.createdBy = properties.createdBy;
    this.credentials = properties.credentials;
    this.id = properties.id;
    this.lastStateError = properties.lastStateError;
    this.modifiedAt = properties.modifiedAt;
    this.modifiedBy = properties.modifiedBy;
    this.name = properties.name;
    this.numberOfEngines = properties.numberOfEngines;
    this.primaryCatalog = properties.primaryCatalog;
    this.projectStore = properties.projectStore;
    this.state = properties.state;
    this.type = properties.type;
    this.#apiMethods = apiMethods;
  }

  async delete() {
    return this.#apiMethods.delete(this.id);
  }

  get settled() {
    return (
      this.state === "ACTIVE" ||
      this.state === "ARCHIVED" ||
      this.state === "INACTIVE"
    );
  }
}
